File: 000001-000500/000202.快乐数/202.js
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(n) {
    let hfactors = {}
    while (n != 1) {
        let sum = 0
        while (n) {
            sum += (n % 10) * (n % 10)
            n = parseInt(n / 10)
        }
        n = sum
        if (hfactors[n]) {
            break
        }
        hfactors[n] = true
    }
    return n == 1
};

console.log(isHappy(9))
console.log(isHappy(19))
console.log(isHappy(2))

File: 000001-000500/000202.快乐数/202.py
class Solution:
    def isHappy(self, n: int) -> bool:
        hfactors = set()
        while (n != 1):
            sum = 0
            while (n):
                sum += (n % 10) * (n % 10)
                n = int(n / 10)
            n = sum
            if n in hfactors:
                break
            hfactors.add(n)
        return n == 1


s = Solution()
print(s.isHappy(9))
print(s.isHappy(19))
print(s.isHappy(2))